src/commands/selectTemplate.ts

```diff
@@ -0,0 +1,107 @@
+};
+
+// NOTE: the default templates are duplicated in package.json, since VSCode offers no way of looking up extension-level default settings
+// So, when modifying them here, be sure to modify them there as well!
+const defaults: { [key in TemplateCommand]: CommandTemplate } = {
+    'build': { label: 'Docker Build', template: 'docker build --rm -f "${dockerfile}" -t ${tag} "${context}"' },
+    'run': { label: 'Docker Run', template: 'docker run --rm -d ${exposedPorts} ${tag}' },
+    'runInteractive': { label: 'Docker Run (Interactive)', template: 'docker run --rm -it ${exposedPorts} ${tag}' },
+    'attach': { label: 'Docker Attach', template: 'docker exec -it ${containerId} ${shellCommand}' },
+    'logs': { label: 'Docker Logs', template: 'docker logs -f ${containerId}' },
+    'composeUp': { label: 'Compose Up', template: 'docker-compose up ${configurationFile} ${detached} ${build}' },
+    'composeDown': { label: 'Compose Down', template: 'docker-compose down ${configurationFile}' },
+};
+
+export async function selectTemplate(context: IActionContext, command: TemplateCommand, matchContext?: string, folder?: vscode.WorkspaceFolder, additionalVariables?: { [key: string]: string }): Promise<string> {
+    // Get the templates from settings
+    const config = vscode.workspace.getConfiguration('docker');
+    const templateSetting: CommandTemplate[] | string = config.get(`commands.${command}`);
+    let templates: CommandTemplate[];
+
+    if (typeof (templateSetting) === 'string') {
+        templates = [{ template: templateSetting }] as CommandTemplate[];
+    } else {
+        templates = templateSetting;
+    }
+
+    // Make sure all templates have some sort of label
+    templates.forEach(template => {
+        template.label = template.label ?? defaults[command].label;
+    })
+
+    // Filter off non-matching / invalid templates
+    templates = templates.filter(template => {
+        if (!template.template) {
+            // Don't wait
+            // eslint-disable-next-line @typescript-eslint/no-floating-promises
+            ext.ui.showWarningMessage(`No command template defined for template '${template.label}'. This template will be skipped.`);
+            return false;
+        }
+
+        if (!template.match) {
+            return true;
+        }
+
+        try {
+            return new RegExp(template.match, 'i').test(matchContext);
+        } catch {
+            // Don't wait
+            // eslint-disable-next-line @typescript-eslint/no-floating-promises
+            ext.ui.showWarningMessage(`Invalid match expression for template '${template.label}'. This template will be skipped.`);
+            return false;
+        }
+    });
+
+    let selectedTemplate: CommandTemplate;
+
+    if (templates.length === 0) {
+        // No templates matched, so fall back to extension default
+        selectedTemplate = defaults[command];
+    } else {
+        selectedTemplate = await quickPickTemplate(context, templates);

```

Comment:
What if the user has templates that are both "matching" as well as "labeled"?  In this implementation, matching templates that don't match aren't considered part of the set of selectable templates.  I can see a scenario where one template has a match property as a kind of "hint" for a well known set of images/containers, but that same template might still be desired for non-matching images/containers.

I think we should attempt to match first and, if no match is found, prompt the user to select a template from the set of all explicitly labeled templates (perhaps also adding the extension/user default).  (I.e. if the user wants a template to *only* be used for a specific match, then she would omit the label property.)

Reply:
I don't think using the label for anything other than display makes sense (i.e. it should not be used for making decisions on whether or not to display a template, users will find that confusing). Also, I think the `match` property _should_ be used to exclude non-matching templates. If they want the same command used for non-matching images, they should define another template with different / no `match` property.

I can see an example where my implementation gives sub-ideal behavior. Suppose the user has two templates, one for MySQL (with a `match`), and one for everything else (without a `match`). If the user right-clicks and runs a MySQL image, they will get prompted to choose between the two templates, even though it's obvious that they would want the MySQL one to be run.

So, I think the right behavior is this:

1. Look for any user-defined matching templates with `match` specified. If multiple, prompt; if none, move to step 2.
2. Look for any user-defined templates with `match` unspecified. If multiple, prompt; if none, move to step 3.
3. Fall back to default (i.e. hard-coded template). (my implementation does this part now)

Reply:
I think that's ok behavior.  Should we just require label properties to help in the case where multiple, matching templates are found (rather than, say, defaulting to the template itself as the label, which won't necessarily make for the nicest looking prompt)?

Reply:
I'm fine with that, I'll make it so.